### PR TITLE
Add inp/outp for the 8080

### PR DESCRIPTION
**Pull request by trondd555**
Uses the same technique as AZTEC C which is to dynamically write the subroutine into reserved data memory and call/jump to it.  This should allow the code to remain ROMable.

While here, change outp_callee to use hl instead of af for stack juggling. af was sending the program off into lala land.

### CODE REVIEW

**Comment by feilipu** [comment]
Maybe fix `wpoke_callee.asm` stack handling while you're at it? Uses `af` too.

**Comment by trondd555** [comment]
Sure.  I'll test and see if it's an issue.  Now that you mentioned it, I see outp.asm does the same thing but I only saw the problem when using af in outp_callee.asm

**Comment by trondd555** [comment]
Oh, outp and wpoke don't get added to the code.  Only *_callee gets added.  That's why I didn't see the issue in there.   I am not sure how these callee functions get chosen so not sure how I could test it.

**Comment by feilipu** [comment]
> Oh, outp and wpoke don't get added to the code. I am not sure how these callee functions get chosen so not sure how I could test it.

The will be called by call by function reference usually, so you could test that way.

But in any case using `af` to hold the return address for an 8080/8085 is an "oops", and it doesn't really need prior testing to warrant fixing it.

**Comment by trondd555** [comment]
Ok, so do you want it fixed just for 8080/8085 or just change it for everything?

**Comment by feilipu** [comment]
> Ok, so do you want it fixed just for 8080/8085 or just change it for everything?

I think there's no problem using `bc` to hold the return addresses for all these `outp.asm`, `wpoke.asm`, and `bpoke.asm` as the general case. So make it so.

We'll get a second opinion before committing.

File: libsrc/stdlib/outp.asm [hunk 1]
```diff
@@ -8,11 +8,12 @@ EXTERN asm_outp
 .outp
 ._outp
 
-   pop af
+   pop bc
```
Comment (feilipu) [inline libsrc/stdlib/outp.asm:11]: Need to use `hl` here.
`bc` is used to hold the port.

Comment (trondd555) [reply libsrc/stdlib/outp.asm]: Oops.  Fixed.  Thanks.

**Comment by feilipu** [comment]
Just thinking, it would be better to use `__CPU_INTEL__` rather than `__CPU_8080__`. The i8085 builds usually include 8080 too, but this would make it clearer.

Sorry to make this late suggestion.

**Comment by suborb** [comment]
Agree with the change to `__CPU_INTEL__`

Couple of other minor things:

- Normally we use Zilog opcodes rather than Intel
- `iotmp` could be shared and 3 bytes would be saved

And as a note `__sfr` access does work on the 808x:

```
__sfr __at 0x20 myport;

int main()
{
        myport = 10;
}
```

**Comment by trondd555** [comment]
Pushed last commit before I saw @suborb's comment.

Working on fixing opcodes.

I don't know where to put iotmp such that it's defined/reserved whenever either function is included.

**Comment by feilipu** [comment]
Use a new `iotmp.asm` file and make the space public. Then reference as `EXTERN`.

**Comment by suborb** [comment]
> I don't know where to put iotmp such that it's defined/reserved whenever either function is included.

The easiest way to do it is with a little bit of a refactor:

- Put the 808x implementations in stdlib/8080
- Add an extra file that contains the "iotmp" variable
- Undo the makefile changes

Sorry for making you go round in circles...

**Comment by feilipu** [comment]
> So is the consensuses to move the code to stdlib/8080?

Yes.

Also then the extra `iotmp.asm` file will go there too, and it won't need `IF __CPU_INTEL__` as its use will be obvious from where it is located.

**Comment by feilipu** [comment]
> Moving the code into stdlib/8080 makes gbz80 pick it up and it fails to find the `EXTERN _iotmp`.

just a minor point, but using a leading underscore allows the variable to be found from C as `iotmp`, and since it is 3 bytes it would not fit into any C type and may conflict with someone's variable. Best not to use a leading underscore in the name, or use two of them to hide it from C.

> Should gbz80 filter out these files from stdlib/8080 now or is there some way to make it work that I don't know?

I'd just wrap the functions and reserved space in `IF __CPU_INTEL__` for the time being, and let @suborb change it later if he likes.

**Comment by trondd555** [comment]
Thanks.

Whoops, I see I accidentally reverted some changes. :(  Thanks for catching those.

**Pull request merged**